src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs

```diff
@@ -14,36 +15,110 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal
             this(duplexPipe, new StreamPipeReaderOptions(leaveOpen: true), new StreamPipeWriterOptions(leaveOpen: true), createStream)
         {
         }
 
-        public DuplexPipeStreamAdapter(IDuplexPipe duplexPipe, StreamPipeReaderOptions readerOptions, StreamPipeWriterOptions writerOptions, Func<Stream, TStream> createStream) : base(duplexPipe.Input, duplexPipe.Output)
+        public DuplexPipeStreamAdapter(IDuplexPipe duplexPipe, StreamPipeReaderOptions readerOptions, StreamPipeWriterOptions writerOptions, Func<Stream, TStream> createStream) :
+            base(duplexPipe.Input, duplexPipe.Output)
         {
             Stream = createStream(this);
+
+            var outputOptions = new PipeOptions(pool: writerOptions.Pool,
+                                                readerScheduler: PipeScheduler.Inline,
+                                                writerScheduler: PipeScheduler.Inline,
+                                                pauseWriterThreshold: 1,
+                                                resumeWriterThreshold: 1,
+                                                minimumSegmentSize: writerOptions.MinimumBufferSize,
+                                                useSynchronizationContext: false);
+
             Input = PipeReader.Create(Stream, readerOptions);
-            Output = PipeWriter.Create(Stream, writerOptions);
+            Output = new Pipe(outputOptions);
         }
 
+        public ILogger Log { get; set; }
+
         public TStream Stream { get; }
 
+        private Pipe Output { get; }
+
         public PipeReader Input { get; }
+        
+        PipeWriter IDuplexPipe.Output
+        {
+            get
+            {
+                if (_outputTask == null)
+                {
+                    _outputTask = WriteOutputAsync();
+                }
 
-        public PipeWriter Output { get; }
+                return Output.Writer;
+            }
+        }
 
-        protected override void Dispose(bool disposing)
+        public override async ValueTask DisposeAsync()
         {
             Input.Complete();
-            Output.Complete();
-            base.Dispose(disposing);
+            Output.Writer.Complete();
+
+            if (_outputTask != null)
+            {
+                // Wait for the output task to complete, this ensures that we've copied
+                // the application data to the underlying stream
+                await _outputTask;
+            }
         }
 
-        public override ValueTask DisposeAsync()
+        private async Task WriteOutputAsync()
         {
-            Input.Complete();
-            Output.Complete();
-            return base.DisposeAsync();
+            try
+            {
+                while (true)
+                {
+                    var result = await Output.Reader.ReadAsync();
+                    var buffer = result.Buffer;
+
+                    try
+                    {
+                        if (buffer.IsEmpty)
```

Comment:
I think you should check result.IsCompleted even if the buffer isn't empty.

Reply:
This code isn't new, why make tweaks to it at this point?

Reply:
It also doesn't make sense to check IsCompleted for non empty buffers. That can result in truncation.